lib/xapi-stdext-date/date.ml

```diff
@@ -30,28 +30,26 @@ let rfc822_to_string x = x
 
 (* ==== ISO8601/RFC3339 ==== *)
 
-(** the name doesn't make much sense anymore, but is kept for compatibility reasons *)
-type iso8601 =
-  | UTC of Ptime.t   (* rfc3339 - accepts only the date in YYYY-MM-DD format *)
-  | Legacy of string (* iso8601 - accepts both YYYYMMDD & YYYY-MM-DD *)
+type iso8601 = Ptime.t
 
  let of_string x =
-  (** prefer to parse with ptime, but rfc3339 does not accept YYYYMMDD)
-    * we fallback on legacy parsing to accept iso8601 datetimes *)
+  let x =
+    try
+      (* if x doesn't contain dashes, insert them, so that ptime can parse x *)
+      Scanf.sscanf x "%04d%02d%02dT%s" (fun y mon d rest ->
+        Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest
+      )
+    with _ -> x
+  in
   match x |> Ptime.of_rfc3339 |> Ptime.rfc3339_error_to_msg with
-  | Error _       -> let assert_utc x =
-                       try Scanf.sscanf x "%_[0-9]T%_[0-9]:%_[0-9]:%_[0-9]Z" ()
```

Comment:
Is assert_utc not needed anymore, because then you would fail below when tz is not 0?

Reply:
yes, we now assert that the input is UTC by checking the timezone 

ptime will additionally accept strings like `"2020-11-11T11:11:11-00:00"`, but i don't see that as an issue, because `"2020-11-11T11:11:11-00:00" |> of_string |> to_string = "20201111T11:11:11Z"` - i.e. we still retain backwards compatibility